sql/core/src/main/scala/org/apache/spark/sql/Observation.scala

```diff
@@ -0,0 +1,150 @@
+
+  @volatile private var row: Option[Row] = None
+
+  /**
+   * Attach this observation to the given [[Dataset]] to observe aggregation expressions.
+   *
+   * @param ds dataset
+   * @param expr first aggregation expression
+   * @param exprs more aggregation expressions
+   * @tparam T dataset type
+   * @return observed dataset
+   * @throws IllegalArgumentException If this is a streaming Dataset (ds.isStreaming == true)
+   */
+  private[spark] def on[T](ds: Dataset[T], expr: Column, exprs: Column*): Dataset[T] = {
+    if (ds.isStreaming) {
+      throw new IllegalArgumentException("Observation does not support streaming Datasets")
+    }
+    register(ds.sparkSession)
+    ds.observe(name, expr, exprs: _*)
+  }
+
+  /**
+   * Get the observed metrics. This waits for the observed dataset to finish its first action.
+   * Only the result of the first action is available. Subsequent actions do not modify the result.
+   *
+   * @return the observed metrics as a [[Row]]
+   * @throws InterruptedException interrupted while waiting
+   */
+  @throws[InterruptedException]
+  def get: Row = {
+    synchronized {
+      // we need to loop as wait might return without us calling notify
+      // https://en.wikipedia.org/w/index.php?title=Spurious_wakeup&oldid=992601610
+      while (this.row.isEmpty) {
+        wait()
+      }
+    }
+
+    this.row.get
+  }
+
+  private def register(sparkSession: SparkSession): Unit = {
+    // makes this class thread-safe:
+    // only the first thread entering this block can set sparkSession
+    // all other threads will see the exception, as it is only allowed to do this once
+    synchronized {
+      if (this.sparkSession.isDefined) {
+        throw new IllegalArgumentException("An Observation can be used with a Dataset only once")
+      }
+      this.sparkSession = Some(sparkSession)
+    }
+
+    sparkSession.listenerManager.register(this.listener)
+  }
+
+  private def unregister(): Unit = {
+    this.sparkSession.foreach(_.listenerManager.unregister(this.listener))
+  }
+
+  private[spark] def onFinish(qe: QueryExecution): Unit = {
+    synchronized {
+      if (this.row.isEmpty) {
+        this.row = qe.observedMetrics.get(name)
+        if (this.row.isDefined) {
+          notifyAll()
+          unregister()
+        }
+      }
+    }
+  }
+
+}
+
+private[sql] case class ObservationListener(observation: Observation)
+  extends QueryExecutionListener {
+
+  override def onSuccess(funcName: String, qe: QueryExecution, durationNs: Long): Unit =
+    observation.onFinish(qe)
+
+  override def onFailure(funcName: String, qe: QueryExecution, exception: Exception): Unit =
+    observation.onFinish(qe)
+
+}
+
+/**
+ * (Scala-specific) Create a named or anonymous instance of Observation.
+ */
+object Observation {
+
+  /**
+   * Observation constructor for creating an anonymous observation.
+   */
+  def apply(): Observation = new Observation(UUID.randomUUID().toString)
```

Comment:
shall we also add a default parameter in the constructor? then java users can also get benefits.

Reply:
Nah, `Observation` won't work anyway, and the default parameter doesn't work in Java side.

Reply:
Oh, we can define another constructor at the class though for Java side.

Reply:
Alright, I have added the unnamed constructor to the Observation class.

I have added a test to `JavaDataFrameSuite` to test the interaction from Java with observations. It shows that the `Dataset.observe` API is not really Java-friendly (see [JavaDataFrameSuite.testObservation](https://github.com/apache/spark/pull/33422/files#diff-5a7751b824f943b4135cbc13a38eab4fee2a27cab7d3fbd9ca331e4a3b638517)), but this shouldn't prevent us from making `Observation` Java-friendly.

The `Dataset.observe` methods could be made Java-friendly (by adding `@varargs`) in a separate PR. @cloud-fan @HyukjinKwon @hvanhovell  What is you opinion on that effort?